common/clock/time_source.go

```diff
@@ -21,51 +21,119 @@
 package clock
 
 import (
-	"sync/atomic"
 	"time"
+
+	"github.com/jonboulle/clockwork"
 )
 
 type (
-	// TimeSource is an interface for any
-	// entity that provides the current
-	// time. Its primarily used to mock
-	// out timesources in unit test
+	// TimeSource provides an interface that packages can use instead of directly using
+	// the [time] module, so that chronology-related behavior can be tested.
 	TimeSource interface {
+		After(d time.Duration) <-chan time.Time
+		Sleep(d time.Duration)
 		Now() time.Time
+		Since(t time.Time) time.Duration
+		NewTicker(d time.Duration) Ticker
+		NewTimer(d time.Duration) Timer
+		AfterFunc(d time.Duration, f func()) Timer
+	}
+
+	// Ticker provides an interface which can be used instead of directly using
+	// [time.Ticker]. The real-time ticker t provides ticks through t.C which
+	// becomes t.Chan() to make this channel requirement definable in this
+	// interface.
+	Ticker interface {
+		Chan() <-chan time.Time
+		Reset(d time.Duration)
+		Stop()
+	}
+
+	// Timer provides an interface which can be used instead of directly using
+	// [time.Timer]. The real-time timer t provides events through t.C which becomes
+	// t.Chan() to make this channel requirement definable in this interface.
+	Timer interface {
+		Chan() <-chan time.Time
+		Reset(d time.Duration) bool
+		Stop() bool
+	}
+
+	// clock serves real wall-clock time
+	clock struct {
+		clockwork.Clock
```

Comment:
huh.  I never really thought about it, but I suppose `&struct{value}` behaves the same (for `value`) as `struct{*value}` since the wrapping-struct essentially compiles to nothing...